### ISSUE TITLE
Facebook Login Issue

### ISSUE THREAD

**Issue by chandrashekharkhod** [abjerner/Skybrud.Social#35]
I've passed all parameters AppID and Secrete Code
Generated AccessToken and redirect url
When user login in facebook and redirect to website it will show error:

// Initialize the Facebook service (no calls are made here)
var service = FacebookService.CreateFromAccessToken(userAccessToken);
var user = service.Users.GetUser(); ******HERE user = null**********

**Comment by abjerner** [comment]
What version of Skybrud.Social are you using? If the Facebook API responds with an error, Skybrud.Social should throw an exception rather than just returning `null`.

Anyways, to test your issue a bit, could you try calling:

```C#
SocialHttpResponse raw = service.Users.Raw.GetUser("me");
```

And then send me the values of both `raw.StatusCode` and `raw.Body`. The latter will contain the raw JSON received from the Facebook API. Notice that if this call is successful, it might have some senstive information about your user, so feel free to blank parts out if you don't want to post them here - I just need to get a general idea of what is going on ;)

**Comment by chandrashekharkhod** [comment]
Thank you for your reply,
But I'm using this in MVCForum. I think all the code is correct by MVCForum.
I've little dough about Facebook Permissions. Should I get Approval from Facebook for my App?
I'm seeing there is "Unapproved" App Info. Please find the screenshot.
Here I just want public data from facebook i.e. emailid, profile picture, Name etc.
![facebook](https://cloud.githubusercontent.com/assets/13394629/8986894/f2356388-36fb-11e5-99ab-7e51aa7e244d.png)

**Comment by abjerner** [comment]
Hi again,

The approved flag is for something else, and shouldn't affect the login, so that is not the culprit.

The intention behind my last post was to see whether the Facebook API returns any errors, and if I haven't handled them correctly in Skybrud.Social. Not knowing this, and since I can't recreate your issue at my end, makes it hard for me to find out whether this is a bug in Skybrud.Social, MCVForum or a wrong configuration somewhere.

**Comment by Rich2020** [comment]
Hi, I am having the same issue. Here is my StatusCode and Body response:

StatusCode: OK
StatusDescription: "OK"
Body: {"name":"Richard Cloete","id":"825993744145433"}
Response.Server is: ""

**Comment by abjerner** [comment]
Hi Richard,

I'm pretty sure your issue is due do something else. Since the JSON received for your user doesn't specify a gender, the parsing will fail, which then will trigger an exception (see #30).

This bug has already been fixed, but I haven't yet had the time to make a new release. Hopefully I'll be able to make that soon ;)

**Comment by Rich2020** [comment]
Hey, thanks for the quick reply.

Awesome, I will wait for the release. Many thanks!

**Comment by abjerner** [comment]
@chandrashekharkhod Given your error report for MvcForum ([link](https://github.com/leen3o/mvcforum/issues/25)), it seems your example is similarly trowing an exception rather than returning null, so I'm assuming you're experiencing the same as Richard (gender not specified in the JSON received from Facebook).

**Comment by Rich2020** [comment]
@abjerner I should also point out that I am developing on my local machine where the site URL is http://localhost:999

Not sure if this will make a difference, but thought I'd say-so, just in case.

**Comment by abjerner** [comment]
@Rich2020 That shouldn't matter, as it is a bug in my parsing ;)

**Comment by Rich2020** [comment]
@abjerner Ahh, okay ;)

**Comment by rlindabury** [comment]
I'm having the same issue with MVCForum using this library. I'm looking forward to the update! :-)

**Comment by abjerner** [comment]
I have now made a new release of Skybrud.Social where the bug should be fixed. To get the new version, simply update the package via NuGet.

I haven't tested in MVCForum though, so please report back whether you still have any issues.

**Comment by rlindabury** [comment]
Fantastic! Perfect timing. We'll be testing this shortly in MVCForum because we're looking to launch tghe forums this week. Thank you!

I'll report back ASAP.

**Comment by rlindabury** [comment]
Ok.. it seems to fail. The message I get on MVC Forum is as follows:

Unable to gain access to your email address - please check your permissions

I'll double-check my FB app but I believe email doesn't require any special perms.

**Comment by abjerner** [comment]
@rlindabury I don't think the email issue has something to do with this. The bug I fixed was that the parsing would fail if the gender wasn't specified in the JSON. For the email, the property should simply be `null` if not present.

Anyways, can you shed some more light in the email error? I think this is more an issue in MVCForum than in Skybrud.Social, as I can't think of anything in my code that would result in that error. The `email` scope is required for the email address to be returned in the JSON from the Graph API.

**Comment by abjerner** [comment]
Hadn't thought of this:

In `v2.4` of the Graph API, only the `name` and `id` properties will be returned by the API unless `fields` is specified in the query string. This is a breaking change from Facebook's side. If you have registered a new app after the launch of `v2.4` of the Graph API, your app will be configured to use this version, and it isn't possible to use older versions. If your app was created before the launch of `v2.4`, it should still work as the email property is exposed by default.

It is my plan to address this in the next minor release of Skybrud.Social (which may be out tomorrow). When fixed in Skybrud.Social, @leen3o probably also need to make a new release of MVCForum.

**Comment by leen3o** [comment]
Could you explain what I would need to change? As I explicitly ask for the email address already? So am a little confused?

https://github.com/leen3o/mvcforum/blob/master/MVCForum.Website/Controllers/OAuthControllers/FacebookOAuthController.cs#L142

Is this not enough in the new apps?

**Comment by abjerner** [comment]
@leen3o With earlier versions of the Graph API, a call to `/me` would return a number of fields by default (including email if the `email` scope is present). Eg. with `v2.3`, a call to `/me` would look like:

```json
{
  "id": "100001100914134",
  "email": "***",
  "first_name": "Anders",
  "gender": "male",
  "last_name": "Bjerner",
  "link": "http://www.facebook.com/100001100914134",
  "locale": "da_DK",
  "name": "Anders Bjerner",
  "timezone": 2,
  "updated_time": "2014-03-08T12:43:18+0000",
  "verified": true
}
```

With `v2.4` only very basic properties are returned by default, so a call to `/me` will only return the ID and the name of the user, not the email address even if the user has granted the `email` scope:

```json
{
  "name": "Anders Bjerner",
  "id": "100001100914134"
}
```

To get the email address (or any other properties for that matter), developers must now specify the `fields` parameter in the query string - eg. by making a call to `/me?fields=id,name,email`.

Since new apps are forced to use `v2.4`, this kinda breaks the logic in Skybrud.Social, since it is build for earlier versions. I'm looking into addressing this with a patch release tonight, so that Skybrud.Social supports specifying a list of fields to be returned.

**Comment by rlindabury** [comment]
@abjerner Thanks for the update and explaination!

I look forward to the new build.

**Comment by abjerner** [comment]
I have now released `v0.9.4.1`, which lets developers specify which fields that should be returned. This means that @leen3o  will have to update MVCForum as well.

A call to get information about the authenticated user could now look like:

```C#
// Declare the options for the call to the API
FacebookGetUserOptions options = new FacebookGetUserOptions {
    Identifier = "me",
    Fields = new[] { "id", "name", "email", "first_name", "last_name", "gender" }
};

// Make the call to the API
FacebookUserResponse response = service.Users.GetUser(options);
```

I'll update the documentation with further information on this when I can find the time ;)

**Comment by leen3o** [comment]
All updated in v1.7 branch. Great work on such a quick fix :+1:

**Comment by abjerner** [comment]
@leen3o Awesome. I suppose Dialogue suffers from the same issue?

Anyways, a little tip: As I'm continuing the development of `v0.9.x`, I try to avoid breaking changes, but then mark a lot of stuff as obsolete. It is my plan to remove most (if not all) of the obsolete code in `v1.0`, so just as a safety measure, I would recommend updating the NuGet dependency for your package to:

```xml
<dependency id="Skybrud.Social" version="[0.9.4.1,1.0.0)" />
```

Thus requiring a version greater than or equal to `v0.9.4.1` (the one released last night), but always lower than `v1.0`.

**Comment by leen3o** [comment]
Yes. I haven't touched Dialogue for a while. Once v1.7 is out for MVCForum then I'm going to jump onto Dialogue.

I have updated the dependency in v1.7 - Thanks.
